R/interpolate.R

```diff
@@ -118,3 +117,121 @@ sqlCommentSpec <- function(start, end, endRequired) {
+  for(c in seq_along(comments)) {
+    comments[[c]][[1]] <- strsplit(comments[[c]][[1]], "", fixed = TRUE)[[1]]
+    comments[[c]][[2]] <- strsplit(comments[[c]][[2]], "", fixed = TRUE)[[1]]
+  }
+  for(q in seq_along(quotes)) {
+    quotes[[q]][[5]] <- nchar(quotes[[q]][[3]]) > 0L
+  }
+
+  state <- 'default'
+  i <- 0
+  while(i < length(sql_arr)) {
+    i <- i + 1
+    switch(state,
+
+      default = {
+        # variable?
+        if (sql_arr[[i]] == "?") {
+          sql_variable_start <- i
+          state <- 'variable'
+          next
+        }
+        # starting quoted area?
+        for(q in seq_along(quotes)) {
+          if (identical(sql_arr[[i]], quotes[[q]][[1]])) {
+            quote_spec_offset <- q
+            state <- 'quote'
+            break
+          }
+        }
+        # we can abort here if the state has changed
+        if (state != 'default') next
+        # starting comment?
+        for(c in seq_along(comments)) {
+          comment_start_arr <- comments[[c]][[1]]
+          comment_start_length <- length(comment_start_arr)
+          if (identical(sql_arr[i:(i + comment_start_length - 1)], comment_start_arr)) {
+            comment_spec_offset <- c
+            i <- i + comment_start_length
+            state <- 'comment'
+            break
+          }
+        }
+      },
+
+      variable = {
+        if (!(sql_arr[[i]] %in% var_chars)) {
+          # make sure variable has at least one character after the '?'
+          if (i - sql_variable_start < 2) {
+            stop("Length 0 variable")
+          }
+          # append current variable offsets to return vectors
+          var_pos_start <- c(var_pos_start, sql_variable_start)
+          var_pos_end <- c(var_pos_end, i - 1)
+          # we have already read too much, go back
+          i <- i - 1
+          state <- 'default'
+        }
+      },
+
+      quote = {
+        # if we see backslash-like escapes, ignore next character
+        if (quotes[[quote_spec_offset]][[5]] && identical(sql_arr[[i]], quotes[[quote_spec_offset]][[3]])) {
+          i <- i + 1
+          next
+        }
+        # end quoted area?
+        if (identical(sql_arr[[i]], quotes[[quote_spec_offset]][[2]])) {
+          quote_spec_offset <- 0L
+          state <- 'default'
+        }
+      },
+
+      comment = {
+        # end commented area?
+        comment_end_arr <- comments[[comment_spec_offset]][[2]]
+        comment_end_length <- length(comment_end_arr)
+        if (identical(sql_arr[i:(i + comment_end_length - 1)], comment_end_arr)) {
+          i <- i + comment_end_length
+          comment_spec_offset <- 0L
+          state <- 'default'
+        }
+      }
+    ) # </switch>
+  } # </while>
+
+  if (quote_spec_offset > 0L) {
+    stop("Unterminated literal")
+  }
+  if (comment_spec_offset > 0L && comments[[comment_spec_offset]][[3]]) {
+    stop("Unterminated comment")
+  }
+  list(start = as.integer(var_pos_start), end = as.integer(var_pos_end))
```

Comment:
If you're careful to keep `i` an integer, the coercion here is probably unnecessary. Again, separate PR.